subprojects/core-api/src/main/java/org/gradle/api/file/RelativePath.java

```diff
@@ -47,17 +48,39 @@ public RelativePath(boolean endsWithFile, String... segments) {
 
     private RelativePath(boolean endsWithFile, @Nullable RelativePath parentPath, String... childSegments) {
         this.endsWithFile = endsWithFile;
-        int targetOffsetForChildSegments;
+        final int expectedLength;
+        final Iterable<String> sourceSegments;
```

Comment:
An alternative here would be to collect a `List<String[]>` containing the `childSegments` and (if present) the `parentPath.segments`. This would add a layer of nested loops, but would allow for a numeric for-loop over the arrays, if desired.